File: app/src/main/java/com/novel/read/fragment/StackFragment.java
package com.novel.read.fragment;

import android.os.Bundle;

import android.widget.TextView;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.mango.mangolib.event.EventManager;
import com.novel.read.R;
import com.novel.read.activity.NovelSearchActivity;
import com.novel.read.adapter.StackAdapter;
import com.novel.read.base.NovelBaseFragment;
import com.novel.read.event.GetCategoryTypeEvent;
import com.novel.read.http.AccountManager;
import com.novel.read.model.protocol.CategoryTypeResp;
import com.novel.read.widget.RefreshLayout;
import com.squareup.otto.Subscribe;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * create by 赵利君 on 2019/6/10
 * describe:
 */
public class StackFragment extends NovelBaseFragment {

    @BindView(R.id.tv_search)
    TextView tvSearch;
    @BindView(R.id.rlv_book_type)
    RecyclerView mRlBookType;
    private StackAdapter mAdapter;
    @BindView(R.id.refresh)
    RefreshLayout refreshLayout;
    private List<CategoryTypeResp.CategoryBean> mList;

    public static StackFragment newInstance() {
        Bundle args = new Bundle();
        StackFragment fragment = new StackFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_stack;
    }

    @Override
    protected void initView() {
        EventManager.Companion.getInstance().registerSubscriber(this);
        mList = new ArrayList<>();
        mRlBookType.setLayoutManager(new GridLayoutManager(getActivity(),2));
        mAdapter = new StackAdapter(mList);
        mRlBookType.setAdapter(mAdapter);

    }

    @Override
    protected void initData() {
        refreshLayout.showLoading();
        getData();
        refreshLayout.setOnReloadingListener(new RefreshLayout.OnReloadingListener() {
            @Override
            public void onReload() {
                getData();
            }
        });
    }

    private void getData(){
        AccountManager.getInstance().getCategoryType();
    }



    @OnClick(R.id.tv_search)
    public void onViewClicked() {
        toActivity(NovelSearchActivity.class);
        getActivity().overridePendingTransition(R.anim.message_fade_in, R.anim.message_fade_out);
    }

    @Subscribe
    public void getCategoryType(GetCategoryTypeEvent event){
        refreshLayout.showFinish();
        if (event.isFail()){
            refreshLayout.showError();
        }else {
            mList.clear();
            mList.addAll(event.getResult().getCategory());
            mAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventManager.Companion.getInstance().unregisterSubscriber(this);
    }
}

File: app/src/main/java/com/novel/read/fragment/StackFragment.kt
package com.novel.read.fragment

import android.os.Bundle
import androidx.recyclerview.widget.GridLayoutManager
import com.mango.mangolib.event.EventManager
import com.novel.read.R
import com.novel.read.activity.NovelSearchActivity
import com.novel.read.adapter.StackAdapter
import com.novel.read.base.NovelBaseFragment
import com.novel.read.event.GetCategoryTypeEvent
import com.novel.read.http.AccountManager
import com.novel.read.model.protocol.CategoryTypeResp
import com.squareup.otto.Subscribe
import kotlinx.android.synthetic.main.fragment_stack.*
import kotlinx.android.synthetic.main.title_stack.*
import java.util.*

/**
 * create by zlj on 2019/6/10
 * describe:
 */
class StackFragment : NovelBaseFragment() {

    private lateinit var mAdapter: StackAdapter
    private var mList: MutableList<CategoryTypeResp.CategoryBean> =ArrayList()

    override fun getLayoutId(): Int {
        return R.layout.fragment_stack
    }

    override fun initView() {
        EventManager.instance.registerSubscriber(this)
        rlv_book_type.layoutManager = GridLayoutManager(activity, 2)
        mAdapter = StackAdapter(mList)
        rlv_book_type.adapter = mAdapter

    }

    override fun initData() {
        refresh.showLoading()
        getData()
        refresh.setOnReloadingListener { getData() }

        tv_search.setOnClickListener {
            toActivity(NovelSearchActivity::class.java)
            activity!!.overridePendingTransition(R.anim.message_fade_in, R.anim.message_fade_out)
        }
    }

    private fun getData() {
        AccountManager.getInstance().getCategoryType()
    }

    @Subscribe
    fun getCategoryType(event: GetCategoryTypeEvent) {
        refresh.showFinish()
        if (event.isFail) {
            refresh.showError()
        } else {
            mList.clear()
            mList.addAll(event.result!!.category)
            mAdapter.notifyDataSetChanged()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        EventManager.instance.unregisterSubscriber(this)
    }

    companion object {
        fun newInstance(): StackFragment {
            val args = Bundle()
            val fragment = StackFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
